### DISCUSSION TITLE
Ordering by JSON field does not cast to number

### DISCUSSION THREAD

**Discussion by bozorgmehr96** [mikro-orm/mikro-orm#4766] [General]
**Describe the bug**
I want to order by a numeric JSON field but it doesn't cast correctly.

**To Reproduce**
```javascript
em.find(Entity, {}, {
orderBy: { jsonField: { numericField: QueryOrder.ASC_NULLS_LAST } },
})
```

It will produce this query:
```sql
select "a".*
from "entity" as "a"
order by "a"."json_field"->>'numericField' asc nulls last
```

**Expected behavior**
The query should be like this:
```sql
select "a".*
from "entity" as "a"
order by ("a"."json_field"->>'numericField')::float8 asc nulls last
```

**Versions**

| Dependency | Version |
| - | - |
| node | v18.15.0 |
| typescript | Version 5.1.6 |
| mikro-orm | 5.7.13 |
| pg | 8.11.1 |

**Comment by B4nan** [comment]
How do you expect this to work? How could the ORM know that your `numericField` is numeric? It's a JSON property, there is no way to tell how it looks from the inside (and in the end, it can contain anything, including mixed data).

You can order by a raw SQL fragment:

https://mikro-orm.io/docs/entity-manager#using-custom-sql-fragments

In v6 this will require `raw` or `sql` helper, but in v5 its just about passing the string in there.

--- 

Or you can use object embeddables, that way you tell the ORM how the shape looks like.

https://mikro-orm.io/docs/embeddables

**Reply by bozorgmehr96** [reply]
My initial solution was to add the functionality by an options field to `orderBy` to cast the type but now that you mention raw queries it doesn't make sense anymore. Why add more complexity when there is query builder?!
Sorry for the rushed issue reporting.

Edit: Thanks for mentioning embeddables, very interesting concept.
